processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -98,287 +167,248 @@ public void start()
   public void stop()
   {
     startStopLock.writeLock().lock();
-    try {
-      if (started.getAndSet(false)) {
-        if (lookupSnapshotTaker != null) {
-          lookupSnapshotTaker.takeSnapshot(getAllAsList());
+
+    if (started) {
+      try {
+        LOG.info("LookupReferencesManager is stopping.");
+        started = false;
+
+        if (exec != null) {
+          exec.shutdownNow();
+          exec = null;
         }
-        LOGGER.info("Stopping lookup factory references manager");
-        for (String lookupName : lookupMap.keySet()) {
-          lookupMap.remove(lookupName).close();
+
+        for (Map.Entry<String, LookupExtractorFactoryContainer> e : lookupMap.entrySet()) {
+          try {
+            LOG.info("Closing lookup [%s]", e.getKey());
+            if (!e.getValue().getLookupExtractorFactory().close()) {
+              LOG.error("Failed to close lookup [%s].");
+            }
+          }
+          catch (Exception ex) {
+            LOG.error(ex, "Failed to close lookup [%s].", e.getKey());
+          }
         }
+
+        lookupMap.clear();
       }
-    }
-    finally {
+      finally {
+        startStopLock.writeLock().unlock();
+      }
+      LOG.info("LookupReferencesManager is stopped.");
+    } else {
       startStopLock.writeLock().unlock();
     }
   }
 
-  /**
-   * @param lookupName             name of the lookupExtractorFactory object
-   * @param lookupExtractorFactory {@link LookupExtractorFactory} implementation reference.
-   *
-   * @return true if the lookup is added otherwise false.
-   *
-   * @throws IllegalStateException If the manager is closed or if start of lookup returns false.
-   */
-  public boolean put(String lookupName, final LookupExtractorFactory lookupExtractorFactory)
+  public void add(String lookupName, LookupExtractorFactoryContainer lookupExtractorFactoryContainer)
   {
+    assertStarted();
```

Comment:
after the stop , it wouldn't add

but there might be a very small period when the element got added and right after that service is stopped ... which is fine because it will stop when node is going down... and anything in the queue dies with that.... when the node comes back up, coordinator would resubmit the request.